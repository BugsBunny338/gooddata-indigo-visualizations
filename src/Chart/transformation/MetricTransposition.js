import { cloneDeep } from 'lodash';
import { enrichHeaders } from './EnrichHeaders';
import { parseValue } from '../../utils/common';

import { DEFAULT_FORMAT } from '../constants/metrics';

function invalidInput(data) {
    return !data || data.isLoading;
}

function transposeData(headers, metrics, rawData) {
    const data = [];

    for (let mi = 0; mi < metrics.length; mi++) {
        for (let ri = 0; ri < rawData.length; ri++) {
            const row = [];

            for (let ci = 0; ci < headers.length; ci++) {
                if (headers[ci].type === 'attrLabel') {
                    const element = rawData[ri][ci];

                    row.push({
                        id: element.id,
                        value: element.name,
                        header: {
                            id: headers[ci].id,
                            uri: headers[ci].uri
                        }
                    });
                }
            }

            const metric = metrics[mi];

            row.push({
                id: metric.header.id, // each metric must have unique id
                value: metric.header.title,
                header: {
                    id: metric.header.id ? metric.header.id : null,
                    uri: metric.header.uri ? metric.header.uri : null
                }
            });

            const metricValue = rawData[ri][metric.index];
            const y = parseValue(metricValue);
            row.push({
                format: metric.header.format || DEFAULT_FORMAT,
                y,
                marker: { enabled: y !== null }
            });

            data.push(row);
        }
    }

    return data;
}

export function transposeMetrics(data) {
    if (invalidInput(data)) {
        return data;
    }

    const dataCopy = cloneDeep(data);

    const { headers, metrics } = enrichHeaders(dataCopy.headers);

    if (metrics.length < 1) return dataCopy;

    return {
        ...dataCopy,
        headers,
        rawData: transposeData(dataCopy.headers, metrics, dataCopy.rawData)
    };
}
