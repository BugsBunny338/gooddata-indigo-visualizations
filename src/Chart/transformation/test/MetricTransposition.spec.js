import { find } from 'lodash';
import { transposeMetrics } from '../MetricTransposition';
import { noMeasure, singleMeasure, singleMeasureAndAttribute } from './fixtures/executionResponses';

describe('MetricTransposition', () => {
    describe('input handling', () => {
        it('should not do anything if data are loading', () => {
            const data = {
                isLoading: true
            };

            expect(transposeMetrics(data)).toEqual(data);
        });

        it('should not do anything if the result is empty', () => {
            const data = null;

            expect(transposeMetrics(data)).toEqual(data);
        });

        it('should not do anything for attributes only', () => {
            const data = noMeasure;

            expect(transposeMetrics(data)).toEqual(data);
        });
    });

    describe('transposition', () => {
        it('should transpose single metric', () => {
            expect(transposeMetrics(singleMeasure)).toMatchSnapshot();
        });

        it('should transpose single metric sliced by attribute', () => {
            expect(transposeMetrics(singleMeasureAndAttribute)).toMatchSnapshot();
        });
    });

    describe('data transformation to metricNames/Values', () => {
        it('should add two extra headers', () => {
            const M1Header = { type: 'metric', title: 'M1', format: '#.##' };
            const M2Header = { type: 'metric', title: 'M2', format: '#.##%' };

            const input = { headers: [M1Header, M2Header], rawData: [] };
            const transformed = transposeMetrics(input);

            const metricNames = find(transformed.headers, { uri: '/metricGroup' });
            const expected = [{ header: M1Header, index: 0 }, { header: M2Header, index: 1 }];
            expect(metricNames.metrics).toEqual(expected);

            const metricValues = find(transformed.headers, { uri: '/metricValues' });
            expect(metricValues.title).toEqual('M1');
        });

        it('should not act on currently loading data', () => {
            expect(transposeMetrics()).toEqual(undefined);

            const data = { isLoading: true };
            expect(transposeMetrics(data)).toEqual(data);
        });

        it('should not do transformation if there are no metrics', () => {
            const data = { headers: [] }; // FIXME: have to inject headers now for _splitHeaders
            expect(transposeMetrics(data)).toEqual(data);
        });
    });
});
