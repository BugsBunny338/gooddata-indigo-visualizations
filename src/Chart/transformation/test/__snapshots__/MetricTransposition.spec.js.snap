// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetricTransposition transposition should transpose single metric 1`] = `
Object {
  "headers": Array [
    Object {
      "id": "metricNames",
      "metrics": Array [
        Object {
          "header": Object {
            "format": "#,##0",
            "id": "a8908b3c92b6743b1fc71c8b113533bc",
            "title": "# of Accounts",
            "type": "metric",
          },
          "index": 0,
        },
      ],
      "title": "Metric",
      "type": "attrLabel",
      "uri": "/metricGroup",
    },
    Object {
      "format": "#,##0",
      "id": "metricValues",
      "title": "# of Accounts",
      "type": "metric",
      "uri": "/metricValues",
    },
  ],
  "rawData": Array [
    Array [
      Object {
        "header": Object {
          "id": "a8908b3c92b6743b1fc71c8b113533bc",
          "uri": null,
        },
        "id": "metric-0",
        "value": "# of Accounts",
      },
      Object {
        "format": "#,##0",
        "marker": Object {
          "enabled": true,
        },
        "y": 118,
      },
    ],
  ],
}
`;

exports[`MetricTransposition transposition should transpose single metric sliced by attribute 1`] = `
Object {
  "headers": Array [
    Object {
      "id": "label.projectgroup.stage",
      "title": "Stage",
      "type": "attrLabel",
      "uri": "/gdc/md/budtwmhq7k94ve7rqj49j3620rzsm3u1/obj/2135",
    },
    Object {
      "id": "metricNames",
      "metrics": Array [
        Object {
          "header": Object {
            "format": "#,##0",
            "id": "a8908b3c92b6743b1fc71c8b113533bc",
            "title": "# of Accounts",
            "type": "metric",
          },
          "index": 1,
        },
      ],
      "title": "Metric",
      "type": "attrLabel",
      "uri": "/metricGroup",
    },
    Object {
      "format": "#,##0",
      "id": "metricValues",
      "title": "# of Accounts",
      "type": "metric",
      "uri": "/metricValues",
    },
  ],
  "rawData": Array [
    Array [
      Object {
        "header": Object {
          "id": "label.projectgroup.stage",
          "uri": "/gdc/md/budtwmhq7k94ve7rqj49j3620rzsm3u1/obj/2135",
        },
        "id": "1",
        "value": "",
      },
      Object {
        "header": Object {
          "id": "a8908b3c92b6743b1fc71c8b113533bc",
          "uri": null,
        },
        "id": "metric-0",
        "value": "# of Accounts",
      },
      Object {
        "format": "#,##0",
        "marker": Object {
          "enabled": true,
        },
        "y": 1,
      },
    ],
    Array [
      Object {
        "header": Object {
          "id": "label.projectgroup.stage",
          "uri": "/gdc/md/budtwmhq7k94ve7rqj49j3620rzsm3u1/obj/2135",
        },
        "id": "321199",
        "value": "Churned",
      },
      Object {
        "header": Object {
          "id": "a8908b3c92b6743b1fc71c8b113533bc",
          "uri": null,
        },
        "id": "metric-0",
        "value": "# of Accounts",
      },
      Object {
        "format": "#,##0",
        "marker": Object {
          "enabled": true,
        },
        "y": 25,
      },
    ],
    Array [
      Object {
        "header": Object {
          "id": "label.projectgroup.stage",
          "uri": "/gdc/md/budtwmhq7k94ve7rqj49j3620rzsm3u1/obj/2135",
        },
        "id": "2288",
        "value": "N/A",
      },
      Object {
        "header": Object {
          "id": "a8908b3c92b6743b1fc71c8b113533bc",
          "uri": null,
        },
        "id": "metric-0",
        "value": "# of Accounts",
      },
      Object {
        "format": "#,##0",
        "marker": Object {
          "enabled": true,
        },
        "y": 1,
      },
    ],
    Array [
      Object {
        "header": Object {
          "id": "label.projectgroup.stage",
          "uri": "/gdc/md/budtwmhq7k94ve7rqj49j3620rzsm3u1/obj/2135",
        },
        "id": "321198",
        "value": "OK",
      },
      Object {
        "header": Object {
          "id": "a8908b3c92b6743b1fc71c8b113533bc",
          "uri": null,
        },
        "id": "metric-0",
        "value": "# of Accounts",
      },
      Object {
        "format": "#,##0",
        "marker": Object {
          "enabled": true,
        },
        "y": 153,
      },
    ],
    Array [
      Object {
        "header": Object {
          "id": "label.projectgroup.stage",
          "uri": "/gdc/md/budtwmhq7k94ve7rqj49j3620rzsm3u1/obj/2135",
        },
        "id": "321216",
        "value": "Potentially Churned / Salesforce Issues",
      },
      Object {
        "header": Object {
          "id": "a8908b3c92b6743b1fc71c8b113533bc",
          "uri": null,
        },
        "id": "metric-0",
        "value": "# of Accounts",
      },
      Object {
        "format": "#,##0",
        "marker": Object {
          "enabled": true,
        },
        "y": 27,
      },
    ],
  ],
}
`;
